### DISCUSSION TITLE
Yarn's 3.1 "pnpm" mode installs all dependencies in the project's node_modules folder

### DISCUSSION THREAD

**Discussion by Thanaen** [yarnpkg/berry#4211] [Q&A]
I wanted to try Yarn's new pnpm node_linker but, as mentionned in the title, it doesn't seem to actually work like pnpm does.
Isn't enabling the pnpm mode supposed to make yarn use a common "store" folder instead of a per-project store?

**Comment by Thanaen** [comment]
Or maybe it's because I hadn't enabled the "enableGlobalCache" setting... 🤔

**Reply by emindeniz99** [reply]
> Or maybe it's because I hadn't enabled the "enableGlobalCache" setting... 🤔

I have enabled it but it works same. pnpm-linker is work per project. I could not see an option that makes it to use global store.

**Reply by Thanaen** [reply]
Yup, looks like pnpm mode is a lie 😁

**Reply by emindeniz99** [reply]
<img width="623" alt="image" src="https://user-images.githubusercontent.com/44745432/159118402-c6d96848-d7bc-4ab3-b953-7596e257bc53.png">


<img width="623" alt="image" src="https://user-images.githubusercontent.com/44745432/159118478-38271946-3be7-4ffb-b2dc-633844563ba7.png">

**Reply by emindeniz99** [reply]
> Yup, looks like pnpm mode is a lie 😁

I tried node_modules linker with hardlinks-global, It links packages with name, multiple version can cause an issue, if I did not understand wrong.
pnpm links with version-hash etc. It seems better to me.

Edit: I can not find any difference between hardlinks-global and hardlinks-local with node_modules linker. 
I am confused.

**Reply by emindeniz99** [reply]
I searched more and found that https://pnpm.io/faq#why-does-my-node_modules-folder-use-disk-space-if-packages-are-stored-in-a-global-store

https://medium.com/@307/hard-links-and-symbolic-links-a-comparison-7f2b56864cdd

:0 hard link is different from symbolic link. I had not known.

**Reply by arcanis** [reply]
> Edit: I can not find any difference between hardlinks-global and hardlinks-local with node_modules linker.

The `hardlinks-global` nm mode hardlinks files from a global store. `hardlinks-local` will use hardlinks if the same file is used multiple times in the same project. The first is more "powerful" but also more susceptible to data corruption.

Our pnpm linker uses symlinks (similar to pnpm), but doesn't implement global file indexing yet.

> Yup, looks like pnpm mode is a lie 😁

Just fyi, any maintainer from any open-source project reading a comment like this is much more likely to ignore support requests.

**Reply by emindeniz99** [reply]
> > Edit: I can not find any difference between hardlinks-global and hardlinks-local with node_modules linker.
> 
> The `hardlinks-global` nm mode hardlinks files from a global store. `hardlinks-local` will use hardlinks if the same file is used multiple times in the same project. The first is more "powerful" but also more susceptible to data corruption.
> 
> Our pnpm linker uses symlinks (similar to pnpm), but doesn't implement global file indexing yet.
> 

Thank you, I learned hard-link.
I also examined `packages/plugin-nm/sources/NodeModulesLinker.ts` in order to understand what happened here. 

I saw that `packages/plugin-pnpm/sources/PnpmLinker.ts` does not have any global file resource as you said.

Thanks for your works.

**Reply by Thanaen** [reply]
> Just fyi, any maintainer from any open-source project reading a comment like this is much more likely to ignore support requests.

Apologies, I didn't want it to sound mean. 😅

**Reply by Thanaen** [reply]
So I wonder: what advantages does the pnpm mode bring? Is it faster or more efficient?

**Comment by MarcCelani-at** [comment]
Are there plans to address the differences between pnpm and the pnpm linker within yarn? The two features I would personally be most interested in are:

1. Hard link to content addressable storage.
2. Global store.

I haven't used it myself, but I believe both of these features are available in the nmMode = `hardlinks-global` config, but that config only appears to be referenced by the classic node modules linker. Is it possible to port this logic over to the pnpm linker?
